hedera-node/docs/design/services/token-service/airdrops/token-claim-airdrop.md

```diff
@@ -0,0 +1,160 @@
+     * This is the ID of the account that SHALL receive the airdrop.<br/>
+     * This field is REQUIRED.
+     */
+    AccountID receiver_id = 2;
+
+    oneof token_reference {
+        /**
+         * A token ID.<br/>
+         * This is the type of token for a fungible/common token airdrop.<br/>
+         * This field is REQUIRED for a fungible/common token and MUST NOT be used for a
+         * non-fungible/unique token.
+         */
+        TokenID fungible_token_type = 3;
+
+        /**
+         * The id of a single NFT, consisting of a Token ID and serial number.<br/>
+         * This is the type of token for a non-fungible/unique token airdrop.<br/>
+         * This field is REQUIRED for a non-fungible/unique token and MUST NOT be used for a
+         * fungible/common token.
+         */
+        NftID non_fungible_token = 4;
+    }
+}
+```
+
+Add new RPC to `TokenService` :
+
+```protobuf
+service TokenService {
+
+//    ...
+
+    /**
+     * Claim one or more pending airdrops.<br/>
+     * This transaction MUST be signed by _each_ account *receiving* an airdrop to be claimed.<br>
+     * If a Sender lacks sufficient balance to fulfill the airdrop at the time the claim is made,
+     * that claim SHALL fail.
+     */
+    rpc claimAirdrop (Transaction) returns (TransactionResponse);
+}
+```
+
+### Fees
+
+The basic `TokenClaimAirdrop` fee should be proportional to the number of airdrops being claimed in the transaction.
+
+Additionally, we should have a fee component for the token association and the crypto transfer that will happen as result of the airdrop claim.
+
+An update into the `feeSchedule` file would be needed to specify that.
+
+### Services updates
+
+- Update `TokenServiceDefinition` class to include the new RPC method definition for claiming airdrops
+- Implement new `TokenClaimAirdropHandler` class which should be invoked when the gRPC server handles `TokenClaimAirdrop` transactions. The class should be responsible for:
+    - Pure checks: validation logic based only on the transaction body itself in order to verify if the transaction is valid one
+        - Verify that the pending airdrops list contains between 1 and 10 entries, inclusive
+        - Verify that the pending airdrops list does not have any duplicate entries
+    - Pre-handle:
+        - The transaction must be signed by the account referenced by a `receiver_id` for each entry in the pending airdrops list
+        - Confirm that for the given pending airdrops ids in the transaction there are corresponding pending transfers existing in state
+        - Check if the sender has sufficient balance to fulfill the airdrop
+    - Handle:
+        - Any additional validation depending on config or state i.e. semantics checks
+        - The business logic for claiming pending airdrops
+            - We need to create a token association between each `receiver_id` and  `token_reference`, future rents for token association slot should be paid by `receiver_id`
```

Comment:
We can explicitly add that since we would have the signature of the receiver, even if it's an account with `receiver_sig_required` property set to true, the claim would implicitly work properly.

Reply:
Added.